File: sdk/dotnet/src/ZoikoTax.Sdk/Result.cs
using System.Diagnostics.CodeAnalysis;

namespace ZoikoTax.Sdk;

/// <summary>
/// The result of a call that returns a value: the value, or the reason there
/// is none.
/// </summary>
/// <remarks>
/// A failed call is a value rather than an exception, because a 4xx is an
/// ordinary outcome a caller is expected to branch on — by
/// <see cref="ZoikoTaxError.ReasonCode"/>, never by message. <see cref="Unwrap"/>
/// is there for callers who prefer exceptions; it is a method on the result
/// rather than a client option, so the client has one behaviour and the choice
/// is visible at the call site.
/// </remarks>
/// <typeparam name="T">The value a successful call produces.</typeparam>
public sealed class Result<T>
{
    internal Result(T? data, ZoikoTaxException? error)
    {
        Data = data;
        Error = error;
    }

    /// <summary>Whether the call succeeded. When true, <see cref="Data"/> is set; when false, <see cref="Error"/> is.</summary>
    [MemberNotNullWhen(true, nameof(Data))]
    [MemberNotNullWhen(false, nameof(Error))]
    public bool IsOk => Error is null;

    /// <summary>The value, when the call succeeded.</summary>
    public T? Data { get; }

    /// <summary>
    /// Why the call failed: a <see cref="ZoikoTaxError"/> when the service
    /// answered with a Problem, a <see cref="ZoikoTaxTransportError"/> when it
    /// did not.
    /// </summary>
    public ZoikoTaxException? Error { get; }

    /// <summary>The value, or the error thrown.</summary>
    /// <example><code>var capabilities = (await client.GetCapabilitiesAsync()).Unwrap();</code></example>
    /// <exception cref="ZoikoTaxError">The service answered with a Problem.</exception>
    /// <exception cref="ZoikoTaxTransportError">The service did not answer.</exception>
    public T Unwrap() => IsOk ? Data : throw Error;
}

/// <summary>
/// The result of a call that returns nothing on success (a <c>204</c>): success,
/// or the reason it failed.
/// </summary>
/// <remarks>
/// A separate type from <see cref="Result{T}"/> rather than a
/// <c>Result&lt;Unit&gt;</c>, so that a caller never has to read a
/// <c>Data</c> that means nothing.
/// </remarks>
public sealed class Result
{
    private static readonly Result Success = new(null);

    private Result(ZoikoTaxException? error)
    {
        Error = error;
    }

    /// <summary>Whether the call succeeded. When false, <see cref="Error"/> is set.</summary>
    [MemberNotNullWhen(false, nameof(Error))]
    public bool IsOk => Error is null;

    /// <summary>
    /// Why the call failed: a <see cref="ZoikoTaxError"/> when the service
    /// answered with a Problem, a <see cref="ZoikoTaxTransportError"/> when it
    /// did not.
    /// </summary>
    public ZoikoTaxException? Error { get; }

    /// <summary>A successful result.</summary>
    public static Result Ok() => Success;

    /// <summary>A failed result.</summary>
    public static Result Fail(ZoikoTaxException error)
    {
        System.ArgumentNullException.ThrowIfNull(error);
        return new(error);
    }

    // The factories for Result<T> live here rather than on it, so a caller
    // writes Result.Ok(value) and the type argument is inferred — which is
    // also what a test double for this client needs.

    /// <summary>A successful result carrying <paramref name="data"/>.</summary>
    public static Result<T> Ok<T>(T data) => new(data, null);

    /// <summary>A failed result for a call that would have returned a <typeparamref name="T"/>.</summary>
    public static Result<T> Fail<T>(ZoikoTaxException error)
    {
        System.ArgumentNullException.ThrowIfNull(error);
        return new(default, error);
    }

    /// <summary>Returns on success, throws the error on failure.</summary>
    /// <exception cref="ZoikoTaxError">The service answered with a Problem.</exception>
    /// <exception cref="ZoikoTaxTransportError">The service did not answer.</exception>
    public void Unwrap()
    {
        if (!IsOk)
        {
            throw Error;
        }
    }
}

File: sdk/java/src/main/java/com/zoikotax/sdk/Result.java
package com.zoikotax.sdk;

import jakarta.annotation.Nonnull;
import jakarta.annotation.Nullable;
import java.util.Objects;

/**
 * The result of a call: the value, or the reason there is none.
 *
 * <p>Every client method returns one of these rather than throwing on a 4xx,
 * because a refused request is an ordinary outcome of calling a tax API and
 * the caller should have to look at it. Sealed, so it can be matched
 * exhaustively:
 *
 * <pre>{@code
 * if (result instanceof Result.Err<Capabilities> err
 *     && err.error() instanceof ZoikoTaxError e) {
 *   log.warn("{} {}", e.getReasonCode(), e.getRequestId());
 * }
 * }</pre>
 *
 * <p>In Kotlin, import it explicitly ({@code import com.zoikotax.sdk.Result});
 * an explicit import takes precedence over the standard library's
 * {@code kotlin.Result}.
 *
 * @param <T> the value's type; {@link Void} for an operation with no body
 */
public sealed interface Result<T> permits Result.Ok, Result.Err {

  /** Whether this is a value rather than a failure. */
  boolean isOk();

  /**
   * The value, or throw the failure.
   *
   * <p>For callers that prefer exceptions, and for tests. It is here rather
   * than a client option, so that the client has one behaviour and the choice
   * is visible at the call site.
   *
   * <pre>{@code
   * Capabilities capabilities = client.getCapabilities().unwrap();
   * }</pre>
   *
   * @throws ZoikoTaxError if the service answered with a Problem
   * @throws ZoikoTaxTransportError if the service was not reached
   */
  T unwrap();

  /** The value, or null if this is a failure (or an operation with no body). */
  @Nullable
  T valueOrNull();

  /** The failure, or null if this is a value. */
  @Nullable
  ZoikoTaxFailure errorOrNull();

  /** A value. For an operation with no response body, {@code value} is null. */
  record Ok<T>(@Nullable T value) implements Result<T> {
    @Override
    public boolean isOk() {
      return true;
    }

    @Override
    public T unwrap() {
      return value;
    }

    @Override
    @Nullable
    public T valueOrNull() {
      return value;
    }

    @Override
    @Nullable
    public ZoikoTaxFailure errorOrNull() {
      return null;
    }
  }

  /** A failure: {@link ZoikoTaxError} or {@link ZoikoTaxTransportError}. */
  record Err<T>(@Nonnull ZoikoTaxFailure error) implements Result<T> {
    public Err {
      Objects.requireNonNull(error, "error");
    }

    @Override
    public boolean isOk() {
      return false;
    }

    @Override
    public T unwrap() {
      throw error;
    }

    @Override
    @Nullable
    public T valueOrNull() {
      return null;
    }

    @Override
    @Nonnull
    public ZoikoTaxFailure errorOrNull() {
      return error;
    }
  }
}
